[UX-335]: RPK list topic regex filter bug

c-julin opened the pull request:
An issue with `rpk topic list` if specified with a regex flag and no matches were found the tool would return all topics.
This issue has been resolved along with some minor optimisations, calling the list topics api only once. Compiling the regex and matching the topics in a single loop. 

New tests have been added which leverages the kfake package introduced in franz-go and runs the command in the test giving us better coverage of the function. A light weight alternative to the python ducktape tests.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=27176) <br/>All committers have signed the CLA.

r-vasquez commented on src/go/rpk/pkg/cli/topic/list_test.go:
```diff
@@ -191,3 +200,180 @@ func TestEmptyTopicList(t *testing.T) {
+
+	var output string
+	// Execute list command and capture output
+	cmd.SetArgs([]string{"list"})
```
What do you think of using `--format` json for this? 

```
	cmd.SetArgs([]string{"list", "--format=json"})
	...
    var o []summarizedList
	json.Unmarshal([]byte(output), &o)
	require.Len(t, o, 3)
	require.Equal(t, "topic1", o[0].Name)
	...
```

That way we can make sure that the length is what we expected, as it is right now we might be printing more rows and won't notice.

c-julin replied:
This is a great suggestion! far more reliable test

r-vasquez commented on src/go/rpk/pkg/cli/topic/list_test.go:
```diff
@@ -191,3 +200,180 @@ func TestEmptyTopicList(t *testing.T) {
+					"Topic %s should NOT be in output for regex %s", topic, tt.regexPattern)
+			}
+
+			// Special check for the bug fix case
+			if tt.name == "no match returns empty" {
+				// Should only contain headers, no actual topics
+				lines := strings.Split(strings.TrimSpace(output), "\n")
+				require.Equal(t, 1, len(lines),
+					"Non-matching regex should only show header line, not all topics")
+				require.Contains(t, lines[0], "NAME")
+				require.Contains(t, lines[0], "PARTITIONS")
+			}
```
Alernatively we could add a function to the test struct and define it in the table. This way if we add another case we won't have to match per name over here.

c-julin replied:
With your above suggestion, we dont need this at all as we test expected output against the json

r-vasquez commented on src/go/rpk/pkg/cli/topic/list_test.go:
```diff
@@ -191,3 +200,180 @@ func TestEmptyTopicList(t *testing.T) {
+}
+
+// captureOutput captures stdout during function execution
+func captureOutput(f func()) string {
```
I do like this, but the only problem I see is that if the code exits with non-zero error code we could stop abruptly the code execution, what do you think of abstracting:

```
			var listedTopics kadm.TopicDetails
			if re {
				listedTopics, err = regexTopicDetails(adm, topics)
				out.MaybeDie(err, "unable to filter topics by regex: %v", err)
			} else {
				listedTopics, err = adm.ListTopicsWithInternal(context.Background(), topics...)
				out.MaybeDie(err, "unable to request metadata: %v", err)
			}

			if detailed {
				printDetailedListView(f, detailedListView(internal, listedTopics), os.Stdout)
			} else {
				printSummarizedListView(f, summarizedListView(internal, listedTopics), os.Stdout)
			}
			out.MaybeDie(err, "unable to summarize metadata: %v", err)
```
			
Into a function that returns an error (to avoid using `out.MaybeDie` outside of run functions) and thest this function instead. This will allow us to test for errors if we want to.

r-vasquez reviewed (commented):
LGTM, some minor comments on the test function but love the first inclusion of kfake in rpk !

c-julin committed:
fix(UX-335): Fixed bug, where all topics would return when regex match
failed. Added cobra tests for list topic using kfake which tests cli
functionality. Tested error scenarios for connection timeout and failed
regex compilations.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d50c1ae3-b808-4421-95b8-7e8a348104c5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70859">build#70859</a></summary>


| test_class                         | test_method                   | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------|:------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest          | test_write                    | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/70859#0198ad8c-1103-4ba9-a107-cd9fa4cb9899 | FLAKY         | 18/21    | upstream reliability is '93.09895833333334'. current run reliability is '85.71428571428571'. drift is 7.38467 and the allowed drift is set to 50. The test should PASS |
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/70859#0198ad8c-10fc-4161-be9f-68cac3fdde87 | FLAKY         | 20/21    | upstream reliability is '98.7603305785124'. current run reliability is '95.23809523809523'. drift is 3.52224 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest           | test_node_operations          | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70859#0198ad8c-10fc-4161-be9f-68cac3fdde87 | FLAKY         | 20/21    | upstream reliability is '99.46236559139786'. current run reliability is '95.23809523809523'. drift is 4.22427 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27176-v25.2.x-3 remotes/upstream/v25.2.x
git cherry-pick -x 5988b3cbb9
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16996596986).

r-vasquez commented:
/backport v25.2.x
